source/client.c

```diff
@@ -491,12 +486,21 @@ int aws_mqtt_client_connection_connect(
 
     connection->host_name = aws_string_new_from_array(connection->allocator, host_name->ptr, host_name->len);
     connection->port = port;
-    connection->tls_options = tls_options;
-    connection->socket_options = socket_options;
+    connection->socket_options = *socket_options;
     connection->state = AWS_MQTT_CLIENT_STATE_CONNECTING;
     connection->clean_session = clean_session;
     connection->keep_alive_time = keep_alive_time;
 
+    /* Cheat and set the tls_options host_name to our copy if they're the same */
+    if (tls_options) {
+        connection->tls_options = *tls_options;
+
+        const char *host_name = (const char *)connection->host_name->bytes;
+        if (0 == strcmp(host_name, connection->tls_options.server_name)) {
```

Comment:
Isn't server_name the result of aws_tls_connection_options_set_server_name()? Is it set by this point? What will happen at clean up time since 2 things reference this stream now?

Reply:
Yes, it is the result of that, and it should be set before this is called. The advantage of keeping tls options by copy now is that we know it can't outlife this connection, and as such, the host_name string can't outlife the tls_options.

Reply:
Do we double free though?

Reply:
WAIT NO HOLD ON, turns out it was being set unconditionally before. I'm going back to that model.

Reply:
Also no, there is no double free, because tls options doesn't free anything.